### PR TITLE
JAPI-499 BinaryRecordReader support U8 to Decimal

### PR DESCRIPTION
**Pull request by jpmcmu**
- Added flag to support treating U8s as BigDecimals
- Created JUnit for U8 to Decimal

Signed-off-by: James McMullan James.McMullan@lexisnexis.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [X] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [ ] The change has been fully tested:
  - [ ] This change does not cause any existing JUnits to fail.
  - [ ] I have include JUnit coverage to test this change
  - [ ] I have performed system test and covered possible regressions and side effects.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] This change fixes the problem, not just the symptom

### CODE REVIEW

**Comment by jpmcmu** [comment]
@drealeed @rpastrana please review

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 1]
```diff
@@ -350,16 +361,23 @@ private Object parseFlatField(FieldDef fd, boolean isLittleEndian) throws Unpars
...
+
+                    if (useDecimalForUnsigned8 && fd.getDataLen() == 8)
+                    {
+                        BigInteger bi = extractUnsigned8Val(intValue);
+                        fieldValue = new BigDecimal(bi);
+                    }
+                    else if (intValue < 0)
                     {
                         messages.addMessage("Warning: Possible unsigned overflow in column: '" + fd.getFieldName() + "'. Convert values to BigInteger via org.hpccsystems.commons.utils.extractUnsigned8 if necessary." );
                     }
```
Comment (drealeed) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:374]: can you mention the setUseDecimalForUnsigned8() method in this message as well, and how it works?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 2]
```diff
@@ -350,16 +361,23 @@ private Object parseFlatField(FieldDef fd, boolean isLittleEndian) throws Unpars
...
                 }
                 else
                 {
                     intValue = getInt((int) fd.getDataLen(), fd.getSourceType() == HpccSrcType.LITTLE_ENDIAN, fd.isBiased());
+                    fieldValue = Long.valueOf(intValue);
                 }
-                fieldValue = Long.valueOf(intValue);
```
Comment (drealeed) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:362]: In ThorFileReader you have

                       if (dataLen == 4 && !isUnsigned) {
                            fieldValue = Integer.valueOf((int) intValue);
                        } else {
                            fieldValue = Long.valueOf(intValue);
                        }

is that check needed here?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 3]
```diff
@@ -807,6 +825,12 @@ private BigDecimal getUnsignedDecimal(int numDigits, int precision, int dataLen)
         return ret;
     }
 
+    private static BigInteger extractUnsigned8Val(long unsigned8)
+    {
+        return (BigInteger.valueOf((unsigned8 >> 32) & 0xffffffffL).shiftLeft(32))
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:830]: what are the expected bitlen and sign of resulting BigInt and how are we ensuring those?
Also, element comment would be helpful.

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/HpccRemoteFileReader.java [hunk 4]
```diff
@@ -240,6 +240,11 @@ public FileReadResumeInfo getFileReadResumeInfo(Long streamPosition)
         return resumeInfo;
     }
 
+    /**
+     * Returns the number of messages create during the reading process
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/HpccRemoteFileReader.java:244]: minor, createD?

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java [hunk 5]
```diff
@@ -334,6 +334,40 @@ public void integrationLargeRecordTest() throws Exception
...
+        // Create a large record dataset
+        FieldDef[] fieldDefs = new FieldDef[1];
+        fieldDefs[0] = new FieldDef("key", FieldType.INTEGER, "UNSIGNED8", 8, true, true, HpccSrcType.LITTLE_ENDIAN, new FieldDef[0]);
+        FieldDef recordDef = new FieldDef("RootRecord", FieldType.RECORD, "rec", 4, false, false, HpccSrcType.LITTLE_ENDIAN, fieldDefs);
+
+        List<HPCCRecord> records = new ArrayList<HPCCRecord>();
+        for (int i = 0; i < 10; i++)
+        {
+            Object[] fields = new Object[1];
+            fields[0] = new BigDecimal(Long.valueOf(i));
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:350]: We should exercise edge cases... Long.MAX_VALUE, etc (take a look at numericOverflowTest)

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java [hunk 6]
```diff
@@ -334,6 +334,40 @@ public void integrationLargeRecordTest() throws Exception
...
+            HPCCRecord record = new HPCCRecord(fields, recordDef);
+            records.add(record);
+        }
+        writeFile(records, "benchmark::unsigned8::10rows", recordDef, connTO);
+
+        HPCCFile file = new HPCCFile("benchmark::unsigned8::10rows", connString , hpccUser, hpccPass);
+        List<HPCCRecord> readRecords = readFile(file, connTO, false, true);
+        if (readRecords.size() < 10)
+        {
+            Assert.fail("Failed to read large record dataset");
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:360]: minor, err message could more accurately describe the issue.

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java [hunk 7]
```diff
@@ -334,6 +334,40 @@ public void integrationLargeRecordTest() throws Exception
...
+    
+    @Test
+    public void unsigned8ToDecimalTest() throws Exception
+    {
+        // Create a large record dataset
+        FieldDef[] fieldDefs = new FieldDef[1];
+        fieldDefs[0] = new FieldDef("key", FieldType.INTEGER, "UNSIGNED8", 8, true, true, HpccSrcType.LITTLE_ENDIAN, new FieldDef[0]);
+        FieldDef recordDef = new FieldDef("RootRecord", FieldType.RECORD, "rec", 4, false, false, HpccSrcType.LITTLE_ENDIAN, fieldDefs);
+
+        List<HPCCRecord> records = new ArrayList<HPCCRecord>();
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:346]: minor, but renaming to orginalRecords would help someone like me understand Line 368 with little effort

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java [hunk 8]
```diff
@@ -701,17 +735,24 @@ public void resumeFileReadTest() throws Exception
             }
         }
     }
-
+
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSReadWriteTest.java:738]: minor, extra whitespace

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java (hunk 2 above)
Comment (jpmcmu) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:362]: @drealeed good question, that check is Tardis specific to match the behavior in the Tardis IRecordWriter. So, we don't need it here.

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java (hunk 3 above)
Comment (jpmcmu) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:830]: Good question, will add some comments here

**Review by drealeed — APPROVED** [review]

**Comment by jpmcmu** [comment]
@rpastrana added additional values to test the entire possible range. Please review

**Review by rpastrana — APPROVED** [review]
